core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java

```diff
@@ -0,0 +1,218 @@
+    int fetched = sourceStream.read(prefixBytes);
+    Preconditions.checkState(fetched == Ciphers.GCM_STREAM_PREFIX_LENGTH,
+        "Insufficient read " + fetched +
+            ". The stream length should be at least " + Ciphers.GCM_STREAM_PREFIX_LENGTH);
+
+    byte[] magic = new byte[Ciphers.GCM_STREAM_MAGIC_ARRAY.length];
+    System.arraycopy(prefixBytes, 0, magic, 0, Ciphers.GCM_STREAM_MAGIC_ARRAY.length);
+    Preconditions.checkState(Arrays.equals(Ciphers.GCM_STREAM_MAGIC_ARRAY, magic),
+        "Cannot open encrypted file, it does not begin with magic string " + Ciphers.GCM_STREAM_MAGIC_STRING);
+
+    if (!emptyCipherStream) {
+      this.plainStreamPosition = 0;
+      this.fileAadPrefix = fileAadPrefix;
+      gcmDecryptor = new Ciphers.AesGcmDecryptor(aesKey);
+      plainBlockSize = ByteBuffer.wrap(prefixBytes, Ciphers.GCM_STREAM_MAGIC_ARRAY.length, 4)
+          .order(ByteOrder.LITTLE_ENDIAN).getInt();
+      Preconditions.checkState(plainBlockSize > 0, "Wrong plainBlockSize " + plainBlockSize);
+
+      cipherBlockSize = plainBlockSize + Ciphers.NONCE_LENGTH + Ciphers.GCM_TAG_LENGTH;
+      this.ciphertextBlockBuffer = new byte[cipherBlockSize];
+      this.currentBlockIndex = 0;
+      this.currentOffsetInPlainBlock = 0;
+
+      int numberOfFullBlocks = Math.toIntExact(netSourceFileSize / cipherBlockSize);
+      int cipherBytesInLastBlock = Math.toIntExact(netSourceFileSize - numberOfFullBlocks * cipherBlockSize);
+      boolean fullBlocksOnly = (0 == cipherBytesInLastBlock);
+      numberOfBlocks = fullBlocksOnly ? numberOfFullBlocks : numberOfFullBlocks + 1;
+      lastCipherBlockSize = fullBlocksOnly ? cipherBlockSize : cipherBytesInLastBlock; // never 0
+      int plainBytesInLastBlock = fullBlocksOnly ? 0 :
+          (cipherBytesInLastBlock - Ciphers.NONCE_LENGTH - Ciphers.GCM_TAG_LENGTH);
+      plainStreamSize = numberOfFullBlocks * plainBlockSize + plainBytesInLastBlock;
+    } else {
+      plainStreamSize = 0;
+
+      gcmDecryptor = null;
+      ciphertextBlockBuffer = null;
+      cipherBlockSize = -1;
+      plainBlockSize = -1;
+      numberOfBlocks = -1;
+      lastCipherBlockSize = -1;
+      this.fileAadPrefix = null;
+    }
+  }
+
+  public long plaintextStreamSize() {
+    return plainStreamSize;
+  }
+
+  @Override
+  public int available() throws IOException {
+    long maxAvailable = plainStreamSize - plainStreamPosition;
+    // See InputStream.available contract
+    if (maxAvailable >= Integer.MAX_VALUE) {
+      return Integer.MAX_VALUE;
+    } else {
+      return (int) maxAvailable;
+    }
+  }
+
+  @Override
+  public int read(byte[] b, int off, int len) throws IOException {
+    if (len < 0) {
+      throw new IOException("Negative read length " + len);
+    }
+
+    if (available() <= 0) {
+      return -1;
+    }
+
+    boolean lastBlock = (currentBlockIndex + 1 == numberOfBlocks);
+    int resultBufferOffset = off;
+    int remaining = len;
```

Comment:
Using `remaining` is a bit confusing because its meaning isn't clear. It could be remaining in the encrypted stream or remaining bytes to read, or something else. I'd recommend being more specific and using `bytesToRead`.

Reply:
SGTM